From: "K. Y. Srinivasan" <kys@microsoft.com>
Date: Tue, 8 Nov 2011 09:01:40 -0800
Subject: [PATCH 03/77] Staging: hv: storvsc: Use mempools to allocate struct
 storvsc_cmd_request

commit 4e03e697c5446146318cdb6344af0060541cbc1c upstream.

We intend to use the storage driver to manage the root device.
To avoid deadlocks, use mempools to allocate struct storvsc_cmd_request.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Signed-off-by: Haiyang Zhang <haiyangz@microsoft.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 drivers/staging/hv/storvsc_drv.c |   30 +++++++++++++++++++++++++-----
 1 file changed, 25 insertions(+), 5 deletions(-)

Index: linux-3.2.46/drivers/staging/hv/storvsc_drv.c
===================================================================
--- linux-3.2.46.orig/drivers/staging/hv/storvsc_drv.c	2013-07-26 18:56:20.000000000 +0000
+++ linux-3.2.46/drivers/staging/hv/storvsc_drv.c	2013-07-26 19:17:59.000000000 +0000
@@ -32,6 +32,7 @@
 #include <linux/module.h>
 #include <linux/device.h>
 #include <linux/hyperv.h>
+#include <linux/mempool.h>
 #include <scsi/scsi.h>
 #include <scsi/scsi_cmnd.h>
 #include <scsi/scsi_host.h>
@@ -42,6 +43,7 @@
 #include <scsi/scsi_dbg.h>
 
 
+#define STORVSC_MIN_BUF_NR				64
 #define STORVSC_RING_BUFFER_SIZE			(20*PAGE_SIZE)
 static int storvsc_ringbuffer_size = STORVSC_RING_BUFFER_SIZE;
 
@@ -287,6 +289,7 @@
 struct hv_host_device {
 	struct hv_device *dev;
 	struct kmem_cache *request_pool;
+	mempool_t *request_mempool;
 	unsigned int port;
 	unsigned char path;
 	unsigned char target;
@@ -975,8 +978,10 @@
 
 	storvsc_dev_remove(dev);
 	if (host_dev->request_pool) {
+		mempool_destroy(host_dev->request_mempool);
 		kmem_cache_destroy(host_dev->request_pool);
 		host_dev->request_pool = NULL;
+		host_dev->request_mempool = NULL;
 	}
 	return 0;
 }
@@ -1126,7 +1131,7 @@
 
 	scsi_done_fn(scmnd);
 
-	kmem_cache_free(host_dev->request_pool, cmd_request);
+	mempool_free(cmd_request, host_dev->request_mempool);
 }
 
 static bool storvsc_check_scsi_cmd(struct scsi_cmnd *scmnd)
@@ -1182,12 +1187,13 @@
 
 	request_size = sizeof(struct storvsc_cmd_request);
 
-	cmd_request = kmem_cache_zalloc(host_dev->request_pool,
+	cmd_request = mempool_alloc(host_dev->request_mempool,
 				       GFP_ATOMIC);
 	if (!cmd_request) {
 		scmnd->scsi_done = NULL;
 		return SCSI_MLQUEUE_DEVICE_BUSY;
 	}
+	memset(cmd_request, 0, sizeof(struct storvsc_cmd_request));
 
 	/* Setup the cmd request */
 	cmd_request->bounce_sgl_count = 0;
@@ -1241,8 +1247,8 @@
 			if (!cmd_request->bounce_sgl) {
 				scmnd->scsi_done = NULL;
 				scmnd->host_scribble = NULL;
-				kmem_cache_free(host_dev->request_pool,
-						cmd_request);
+				mempool_free(cmd_request,
+						host_dev->request_mempool);
 
 				return SCSI_MLQUEUE_HOST_BUSY;
 			}
@@ -1284,7 +1290,7 @@
 			destroy_bounce_buffer(cmd_request->bounce_sgl,
 					cmd_request->bounce_sgl_count);
 
-		kmem_cache_free(host_dev->request_pool, cmd_request);
+		mempool_free(cmd_request, host_dev->request_mempool);
 
 		scmnd->scsi_done = NULL;
 		scmnd->host_scribble = NULL;
@@ -1354,6 +1360,7 @@
 			const struct hv_vmbus_device_id *dev_id)
 {
 	int ret;
+	int number = STORVSC_MIN_BUF_NR;
 	struct Scsi_Host *host;
 	struct hv_host_device *host_dev;
 	bool dev_is_ide = ((dev_id->driver_data == IDE_GUID) ? true : false);
@@ -1382,8 +1389,19 @@
 		return -ENOMEM;
 	}
 
+	host_dev->request_mempool = mempool_create(number, mempool_alloc_slab,
+						mempool_free_slab,
+						host_dev->request_pool);
+
+	if (!host_dev->request_mempool) {
+		kmem_cache_destroy(host_dev->request_pool);
+		scsi_host_put(host);
+		return -ENOMEM;
+	}
+
 	stor_device = kzalloc(sizeof(struct storvsc_device), GFP_KERNEL);
 	if (!stor_device) {
+		mempool_destroy(host_dev->request_mempool);
 		kmem_cache_destroy(host_dev->request_pool);
 		scsi_host_put(host);
 		return -ENOMEM;
@@ -1398,6 +1416,7 @@
 	stor_device->port_number = host->host_no;
 	ret = storvsc_connect_to_vsp(device, storvsc_ringbuffer_size);
 	if (ret) {
+		mempool_destroy(host_dev->request_mempool);
 		kmem_cache_destroy(host_dev->request_pool);
 		scsi_host_put(host);
 		kfree(stor_device);
@@ -1437,6 +1456,7 @@
 
 err_out:
 	storvsc_dev_remove(device);
+	mempool_destroy(host_dev->request_mempool);
 	kmem_cache_destroy(host_dev->request_pool);
 	scsi_host_put(host);
 	return -ENODEV;
